packages/svelte/src/compiler/phases/utils.js

```diff
@@ -0,0 +1,41 @@
+/** @import { AST, Binding } from '#compiler' */
+/** @import { Scope } from './scope' */
+
+/**
+ * Whether a variable can be referenced directly from template string.
+ * @param {Binding | undefined} binding
+ * @returns {boolean}
+ */
+function can_inline_variable(binding) {
+	return (
+		!!binding &&
+		// in a `<script module>` block
+		!binding.scope.parent &&
+		// to prevent the need for escaping
+		binding.initial?.type === 'Literal'
+	);
+}
+
+/**
+ * @param {AST.Attribute} attribute
+ * @param {Scope} scope
+ */
+export function is_inlinable_expression(attribute, scope) {
+	if (attribute.value === true) return false; // not an expression
+	let nodes = Array.isArray(attribute.value) ? attribute.value : [attribute.value];
+	let has_expression_tag = false;
+	for (let value of nodes) {
+		if (value.type === 'ExpressionTag') {
+			if (value.expression.type === 'Identifier') {
```

Comment:
the more i look at this function the less i understand it. why is this the test? it excludes lots of things that absolutely _could_ be inlined

Reply:
I think the initial inline work was made very strict on purpose to get the immediate gain of an easy case (const variables inlined in attributes) without the complexity of more complex cases.